### PR TITLE
Fix non-tensor values ​​in batch used for pipeline parallelization

### PR DESCRIPTION
**Pull request by yhna940**
# Motivation

Fix bugs in pipeline parallelization to support non-tensor values ​​in batch.

# Modification

1.  In `Copy` autograd function, add logic to check if it is a tensor.
2. Changed to check whether it is a tensor first in the `profile` process for layer partitioning.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### CODE REVIEW

**Comment by kit1980** [comment]
@awgu Could you take a look?

File: torch/distributed/pipeline/sync/copy.py [hunk 1]
```diff
@@ -65,14 +65,17 @@ def backward(ctx: Context, *grad_output: Tensor,) -> Tuple[Optional[Tensor], ...
...
-                # 'next_stream' is not where 'x' has been allocated.
-                record_stream(x, next_stream)
-                # 'y' has been allocated on 'prev_stream'.
-                # It might be used on the current stream captured as 'input_stream'.
-                record_stream(y, input_stream)
+                if torch.is_tensor(x):
+                    y = x.to(get_device(prev_stream), non_blocking=True)
+                    grad_input.appendleft(y)
+
+                    # 'next_stream' is not where 'x' has been allocated.
```
Comment (awgu) [inline torch/distributed/pipeline/sync/copy.py:72]: nit: Should we use backtick \` instead of single quote \'?

**Review by awgu — COMMENTED** [review]
This looks good to me overall, but I wonder if it would be possible to write a unit test for this.

**Comment by KKIEEK** [comment]
I think these lines also need modification.
https://github.com/pytorch/pytorch/blob/59aacc40ca2248a18af385cd30831ee785bbb684/torch/distributed/pipeline/sync/_balance/profile.py#L94-L96

File: torch/distributed/pipeline/sync/_balance/profile.py [hunk 2]
```diff
@@ -91,9 +96,12 @@ def profile_sizes(
     batch = Batch(input)
     sizes: List[int] = []
 
-    latent_scale = batch[0].size(0) / chunks
+    tensor_idx = batch.find_tensor_idx()
+    latent_scale = batch[tensor_idx].size(0) // chunks
```
Comment (awgu) [inline torch/distributed/pipeline/sync/_balance/profile.py:100]: I am not familiar enough with this code, so I just wanted to confirm you want the truncation behavior with this integer division.

File: torch/distributed/pipeline/sync/_balance/profile.py [hunk 3]
```diff
@@ -38,16 +38,21 @@ def layerwise_sandbox(module: nn.Sequential, device: torch.device,) -> Generator
 def detach(batch: Batch) -> None:
     """Detaches from autograd graph."""
     for i, x in enumerate(batch):
+        if not torch.is_tensor(x):
+            continue
         batch[i] = x.detach().requires_grad_(x.requires_grad)
 
 
+
```
Comment (awgu) [inline torch/distributed/pipeline/sync/_balance/profile.py:46]: nit: Should we keep 2 newlines between functions? (i.e. delete this added newline)

File: torch/distributed/pipeline/sync/_balance/profile.py [hunk 4]
```diff
@@ -108,7 +116,7 @@ def profile_sizes(
         param_size = sum(p.storage().nbytes() for p in layer.parameters())
 
         # Combine size of parameters and activations with normalize scales.
-        size = latent_size * latent_scale + param_size * param_scale
+        size = latent_size * chunks + param_size * param_scale
```
Comment (awgu) [inline torch/distributed/pipeline/sync/_balance/profile.py:119]: Same here with this `size` calculation -- just want to make sure this is intended

**Review by awgu — COMMENTED** [review]
Thanks for continuing to work on this! I left some more comments. I am not the code owner of this pipeline implementation and am not that familiar, so let me see if we can get a more informed review.

@mrshenli Do you know who might be able to review this more deeply?

**Comment by awgu** [comment]
Could we also add a PR description? Thanks!

File: torch/distributed/pipeline/sync/_balance/profile.py (hunk 2 above)
Comment (yhna940) [reply torch/distributed/pipeline/sync/_balance/profile.py]: There was a mistake on my part and I corrected it. (Like the original, set the micro-batch size to 1, pass it through the pipeline, and return to the actual micro-batch size to be used.) Thank you very much for confirming again. However, it seems that `find_tensor_idx` should still be used because the order of tensors in a batch cannot be guaranteed.

File: torch/distributed/pipeline/sync/_balance/profile.py (hunk 4 above)
Comment (yhna940) [reply torch/distributed/pipeline/sync/_balance/profile.py]: As mentioned above, it was my mistake. restored to original Thank you very much.

File: test/distributed/pipeline/sync/test_balance.py [hunk 5]
```diff
@@ -223,3 +223,32 @@ def test_already_has_grad():
 
     with pytest.raises(ValueError, match="some parameter already has gradient"):
         balance_by_time(1, model, sample, device="cpu")
+
+
```
[the remark this answers is not in the document]
Comment (yhna940) [reply test/distributed/pipeline/sync/test_balance.py]: As you pointed out, I added an assertion for non-tensors (`test_balance_by_size_tuple`). Instead, in the case of profiling time, it is simplified by passing non-tensors in `test_balance_by_time`. It seems difficult to add assertions for non-tensors. 😅 Are there any more unit tests I should add? thank you

**Comment by yhna940** [comment]
> Looks like this PR hasn't been updated in a while so we're going to go ahead and mark this as `Stale`. 
> Feel free to remove the `Stale` label if you feel this was a mistake. 
> If you are unable to remove the `Stale` label please contact a maintainer in order to do so. 
> If you want the bot to never mark this PR stale again, add the `no-stale` label.
> `Stale` pull requests will automatically be closed after 30 days of inactivity.

I really need this feature, can you check it? thank you :)
